### DISCUSSION TITLE
On Windows, exclusive fullscreen only activates on OpenGL

### DISCUSSION THREAD

**Discussion by jvyden** [ppy/osu#30349] [Q&A]
### Type

Performance

### Bug description

I recently switched from Linux on X11 to Windows 11, and immediately upon opening the game I noticed something was off; I could feel a *ton* of latency.

After some experimentation with settings I've determined that osu! doesn't seem to be entering exclusive fullscreen mode on renderers other than OpenGL.

Testing procedure was as follows:
1. Open the game on X renderer
2. Attempt to change the system volume so Windows tries displays a toast over the game; if it's shown EFS is not working.
4. Open settings and move the mouse around to feel for latency
5. Export logs

### Screenshots or videos

Not sure what I could provide here that would be helpful.

### Version

2024.1009.1-lazer

### Logs

OpenGL (working!): 
[compressed-logs.zip](https://github.com/user-attachments/files/17441598/compressed-logs.zip)
Direct3D 11 (EFS does not work): 
[compressed-logs.zip](https://github.com/user-attachments/files/17441603/compressed-logs.zip)
Direct3D 11 (Experimental) (EFS does not work): 
[compressed-logs.zip](https://github.com/user-attachments/files/17441604/compressed-logs.zip)
Vulkan (Experimental) (EFS does not work): 
[compressed-logs.zip](https://github.com/user-attachments/files/17441606/compressed-logs.zip)

DxDiag: [DxDiag.txt](https://github.com/user-attachments/files/17441649/DxDiag.txt)

**Comment by peppy** [comment]
direct x 11 doesn't use (or need) exclusive full screen for lower latency. If you're seeing latency you may need to update device drivers.

**Reply by jvyden** [reply]
I was on a pretty recent driver from Windows Update (NVIDIA 560.XX) but I updated anyways to 565.90 and I'm still seeing the same behavior. It 100% feels like there's a frame or two of latency and it feels pretty terrible since I'm on a 60hz monitor.

**Reply by Espionage724** [reply]
Dota 2 is 20+ FPS difference in real 720p vs 1080p scaled to 720p, while doing the exclusive fullscreen res/refresh rate change. I'm not sure what Microsoft might say about latency, but I get better performance on real lower resolutions.

___

I haven't tested it yet, but got a new display and wanted to test it at my usual 1024x768. `2025.912.0-lazer` any option D3D11/VK/OGL at the Fullscreen option keeps the native 1080p res and does some scaling to the lower resolutions.

**Comment by SupDos** [comment]
Just as an extra note, I had a slightly similar kerfuffle with exclusive fullscreen, but in my case it stopped working in OpenGL after a GPU driver update (https://github.com/ppy/osu/discussions/29647). TL;DR: in *Hardware: Independent Flip* mode, you'll still see the volume controls show up but it is essentially (and I had to just accept this) the same as Hardware: Legacy Flip (EFS) in terms of latency.

It might be worth checking using PresentMon what the game is running in using the other renderers. If it's anything other than *Hardware: Legacy Flip* or *Hardware: Independent Flip* then it's not right

**Reply by jvyden** [reply]
Yeah, I'm seeing `Composed: Flip` on Direct3D, interestingly. On OpenGL I get `Hardware: Legacy Flip` as I would expect.

**Reply by sz-da** [reply]
> Yeah, I'm seeing `Composed: Flip` on Direct3D, interestingly. On OpenGL I get `Hardware: Legacy Flip` as I would expect.

Seeing the same thing here on Windows 10 & 11, though this only happens when the game is running on the non-primary monitor.
When running on the main monitor, Direct3D runs in `Hardware: Independent Flip` mode as expected.

**Reply by peppy** [reply]
cc @smoogipoo for visibility
